chain/client/src/archive/cold_store_actor.rs

```diff
@@ -0,0 +1,496 @@
+        )?;
+
+        if let Some(cloud_storage) = &self.cloud_storage {
+            update_cloud_storage(cloud_storage, &self.hot_store, &next_height)?;
+        }
+
+        update_cold_head(&self.cold_db, &self.hot_store, &next_height)?;
+
+        let result = if next_height >= hot_final_head_height {
+            Ok(ColdStoreCopyResult::LatestBlockCopied)
+        } else {
+            Ok(ColdStoreCopyResult::OtherBlockCopied)
+        };
+
+        tracing::trace!(target: "cold_store", ?result, "ending");
+        result
+    }
+}
+
+// Check some basic sanity conditions.
+// * cold head <= hot final head
+// * cold head >= hot tail
+fn sanity_check_impl(
+    cold_head_height: u64,
+    hot_final_head_height: u64,
+    hot_tail_height: u64,
+) -> anyhow::Result<(), ColdStoreError> {
+    // We should only copy final blocks to cold storage.
+    if cold_head_height > hot_final_head_height {
+        return Err(ColdStoreError::ColdHeadAheadOfFinalHeadError {
+            cold_head_height,
+            hot_final_head_height,
+        });
+    }
+
+    // Cold and Hot storages need to overlap. Without this check we would skip
+    // blocks from cold_head_height to hot_tail_height. This will result in
+    // corrupted cold storage.
+    if cold_head_height < hot_tail_height {
+        return Err(ColdStoreError::ColdHeadBehindHotTailError {
+            cold_head_height,
+            hot_tail_height,
+        });
+    }
+    Ok(())
+}
+
+fn cold_store_copy_result_to_string(
+    result: &anyhow::Result<ColdStoreCopyResult, ColdStoreError>,
+) -> &str {
+    match result {
+        Err(ColdStoreError::ColdHeadBehindHotTailError { .. }) => "cold_head_behind_hot_tail_error",
+        Err(ColdStoreError::ColdHeadAheadOfFinalHeadError { .. }) => {
+            "cold_head_ahead_of_final_head_error"
+        }
+        Err(ColdStoreError::SkippedBlocksBetweenColdHeadAndNextHeightError { .. }) => {
+            "skipped_blocks_between_cold_head_and_next_height_error"
+        }
+        Err(ColdStoreError::ColdHeadHashReadError { .. }) => "cold_head_hash_read_error",
+        Err(ColdStoreError::EpochError { .. }) => "epoch_error",
+        Err(ColdStoreError::Error { .. }) => "error",
+        Ok(ColdStoreCopyResult::NoBlockCopied) => "no_block_copied",
+        Ok(ColdStoreCopyResult::LatestBlockCopied) => "latest_block_copied",
+        Ok(ColdStoreCopyResult::OtherBlockCopied) => "other_block_copied",
+    }
+}
+
+#[derive(Debug)]
+enum ColdStoreMigrationResult {
+    /// Cold storage was already initialized
+    NoNeedForMigration,
+    /// Performed a successful cold storage migration
+    SuccessfulMigration,
+    /// Migration was interrupted by keep_going flag
+    MigrationInterrupted,
+}
+
+pub struct ColdStoreActor {
+    split_storage_config: SplitStorageConfig,
+    genesis_height: BlockHeight,
+    hot_store: Store,
+    cold_db: Arc<ColdDB>,
+    cloud_storage: Option<Arc<CloudStorage>>,
```

Comment:
I'm mildly annoyed that they are called differently (store, storage and db). Nothing to do in this PR.